mne/channels/channels.py

```diff
@@ -740,6 +740,32 @@ def pick_types(self, meg=None, eeg=False, stim=False, eog=False,
             ias=ias, syst=syst, seeg=seeg, dipole=dipole, gof=gof, bio=bio,
             ecog=ecog, fnirs=fnirs, include=include, exclude=exclude,
             selection=selection)
+
+        # remove dropped channel types from reject and flat
+        if getattr(self, 'reject', None) is not None:
+            if (meg is False or meg == 'mag') and 'grad' in self.reject:
+                del self.reject['grad']
+            if (meg is False or meg == 'grad') and 'mag' in self.reject:
+                del self.reject['mag']
+            if not eeg and 'eeg' in self.reject:
+                del self.reject['eeg']
+            if not eog and 'eog' in self.reject:
+                del self.reject['eog']
+            if not ecg and 'ecg' in self.reject:
+                del self.reject['ecg']
+
+        if getattr(self, 'flat', None) is not None:
+            if (meg is False or meg == 'mag') and 'grad' in self.flat:
+                del self.flat['grad']
+            if (meg is False or meg == 'grad') and 'mag' in self.flat:
+                del self.flat['mag']
+            if not eeg and 'eeg' in self.flat:
+                del self.flat['eeg']
+            if not eog and 'eog' in self.flat:
+                del self.flat['eog']
+            if not ecg and 'ecg' in self.flat:
+                del self.flat['ecg']
```

Comment:
this solution will is not generic enough. You can have seeg, fnirs, ecog channnels etc. One solution I would suggest:

you pick and get the new info. Then you loop over the reject dict and you can use the "in" syntax.

```
for ch_type in reject:
    if ch_type not in self:
          del self.reject['ch_type]
```

Reply:
Didn't realize the "in" syntax was an option! Made changes to match what you suggested - had to make a few other edits to get things working, including moving the final call to `self._pick_drop_channels` (hopefully that's okay).